### PR TITLE
Implement hardswish/hardsigmoid on MKLDNN tensors

### PR DESCRIPTION
**Pull request by Krovatkin**
Adding hardwish and hardsigmoid improves mobilenetv3 by ~13%

  | hardswish | base |  
-- | -- | -- | --
run 1 | 1305.032 | 1486.013 |  
run 2 | 1290.142 | 1491.001 |  
run 3 | 1305.51 | 1491.66 |  
run 4 | 1308.788 | 1495.577 |  
avg | 1302.368 | 1491.063 | 0.873449

### CODE REVIEW

File: torch/csrc/jit/passes/frozen_ops_to_mkldnn.cpp [hunk 1]
```diff
@@ -205,6 +213,41 @@ void InplaceMKLDNNSubgraph(std::shared_ptr<Graph> graph) {
   }
 }
 
+Operation createUnaryOp(
+    std::function<void(at::Tensor output, at::Tensor input)> aten_op,
```
Comment (eellison) [inline torch/csrc/jit/passes/frozen_ops_to_mkldnn.cpp:217]: we are never running the op in Aten. it is invariant that if we fuse the op into the mkldnn subgraph all inputs are mkldnn tensors

File: torch/csrc/jit/passes/frozen_ops_to_mkldnn.cpp [hunk 2]
```diff
@@ -205,6 +213,41 @@ void InplaceMKLDNNSubgraph(std::shared_ptr<Graph> graph) {
...
+    std::function<Tensor(Tensor)> fallback) {
+  return [aten_op, fallback](Stack* stack) {
+    auto a = pop(stack).toTensor();
+    if (a.numel() == 0) {
```
Comment (eellison) [inline torch/csrc/jit/passes/frozen_ops_to_mkldnn.cpp:221]: if this is just to put numel() == 0 on the hot path instead of for correctness reasons is it worth it ?

File: torch/csrc/jit/passes/frozen_ops_to_mkldnn.cpp [hunk 3]
```diff
@@ -633,6 +692,10 @@ class MKLDNNSubgraphSlicer {
         return true;
     }
 
+    if (n->kind() == Symbol::aten("hardswish")) {
```
Comment (eellison) [inline torch/csrc/jit/passes/frozen_ops_to_mkldnn.cpp:695]: nit: register `aten::hardswish as an interned operator ` ?

File: torch/csrc/jit/passes/frozen_ops_to_mkldnn.cpp [hunk 4]
```diff
@@ -205,6 +213,41 @@ void InplaceMKLDNNSubgraph(std::shared_ptr<Graph> graph) {
...
+      push(stack, a.clone());
+    }
+    if (a.is_mkldnn()) {
+      c10::impl::ExcludeDispatchKeyGuard edkg(c10::autograd_dispatch_keyset);
+      auto a_it = at::native::itensor_from_mkldnn(a);
+      auto raw_data = a_it.get_data_handle();
+      auto topt = a.options().layout(c10::kStrided);
+      auto t = at::from_blob(raw_data, {a.numel()}, topt);
+
+      auto it_empty = ideep::tensor(a_it.get_desc());
```
Comment (eellison) [inline torch/csrc/jit/passes/frozen_ops_to_mkldnn.cpp:232]: im not really familiar with this code but have we verified that we're doing what we're think we're doing ? could you add a comment at the top of this block explaining whats happening ?

File: torch/csrc/jit/passes/frozen_ops_to_mkldnn.cpp [hunk 5]
```diff
@@ -456,6 +510,11 @@ void ComputeSubgraphInMKLDNN(Node* subgraph_node) {
       body_node->replaceInput(1, node->outputs().at(1));
     }
 
+    if (body_node->kind() == Symbol::aten("hardswish")) {
+      body_node->replaceWithNewSymbol(Symbol::aten("MKLDNNHardSwish"));
+      body_node->destroy();
```
Comment (eellison) [inline torch/csrc/jit/passes/frozen_ops_to_mkldnn.cpp:515]: We should be adding `hardswish_` here https://github.com/pytorch/pytorch/blob/ce32d3d54b4bad74ecc4b44f68b1ca94b734e7ef/torch/csrc/jit/passes/frozen_ops_to_mkldnn.cpp#L819

File: torch/csrc/jit/passes/frozen_ops_to_mkldnn.cpp [hunk 6]
```diff
@@ -238,6 +281,17 @@ Operation BroadOp(const Node* node) {
   };
 }
 
+const RegisterOperators MKLDNNHardSwishOpReg({
+    torch::jit::Operator(
+        "aten::MKLDNNHardSwish(Tensor a) -> Tensor",
```
Comment (eellison) [inline torch/csrc/jit/passes/frozen_ops_to_mkldnn.cpp:286]: We should have a way of inplacing this : https://github.com/pytorch/pytorch/blob/ce32d3d54b4bad74ecc4b44f68b1ca94b734e7ef/torch/csrc/jit/passes/frozen_ops_to_mkldnn.cpp#L180.

Maybe just by adding an `inplace=true` attribute to the node

File: torch/csrc/jit/passes/frozen_ops_to_mkldnn.cpp [hunk 7]
```diff
@@ -238,6 +281,17 @@ Operation BroadOp(const Node* node) {
   };
 }
 
+const RegisterOperators MKLDNNHardSwishOpReg({
+    torch::jit::Operator(
```
Comment (eellison) [inline torch/csrc/jit/passes/frozen_ops_to_mkldnn.cpp:285]: nit: we have `prim::mkldnn_convolution`, reason not to put this in as `prim::hardswish` for consistency ?

**Review by eellison — COMMENTED** [review]
Looks good!

File: test/jit/test_freezing.py [hunk 8]
```diff
@@ -1827,6 +1829,34 @@ def forward(self, x):
...
+                        graph_str = f"""graph(%input.1 : Tensor):
+                            %33 : None = prim::Constant()
+                            %34 : Tensor = aten::to_mkldnn(%input.1, %33)
+                            %35 : Tensor = {mkldnn_opname}{inplace_str}(%34)
+                            return ({inplace_tgt})
+                        """
+                        g = parse_ir(graph_str)
+                        m = self.createFunctionFromGraph(g)
+                        x = torch.rand(size)
+                        x_copy = x.detach().clone()
```
Comment (eellison) [inline test/jit/test_freezing.py:1856]: nit: why are you detaching and cloning x_copy ?

File: test/jit/test_freezing.py [hunk 9]
```diff
@@ -1853,9 +1883,11 @@ def checkResults(self, mod1, mod2):
 
     def test_successful(self):
         # simple conv-relu
-        mod_eager = nn.Sequential(self.getConv(), nn.ReLU(), nn.ReLU())
+
+        mod_eager = nn.Sequential(self.getConv(), nn.Hardswish(), nn.ReLU())
         mod = self.freezeAndConvert(mod_eager)
-        FileCheck().check("mkldnn_convolution").check_next("aten::relu_").check_next("aten::relu_").run(mod.graph)
+        print(mod.graph)
```
Comment (eellison) [inline test/jit/test_freezing.py:1889]: stray print

File: test/jit/test_freezing.py [hunk 10]
```diff
@@ -1853,9 +1883,11 @@ def checkResults(self, mod1, mod2):
...
+        FileCheck().check("mkldnn_convolution").check_next("prim::MKLDNNHardSwish_").check_next("aten::relu_").run(mod.graph)
```
Comment (eellison) [inline test/jit/test_freezing.py:1890]: is the hardswish still necessary

**Review by eellison — APPROVED** [review]
Nice, LGTM!

File: test/jit/test_freezing.py (hunk 9 above)
Comment (Krovatkin) [reply test/jit/test_freezing.py:1885]: a great catch!

File: test/jit/test_freezing.py (hunk 8 above)
Comment (Krovatkin) [reply test/jit/test_freezing.py:1853]: for in-place tests, the reference implementations modify the input in `aten_op(x, inplace=inplace)`. I guess, I could just always use  `F.hardswish(inplace=False)` , might look a bit cleaner.

File: torch/csrc/jit/passes/frozen_ops_to_mkldnn.cpp [hunk 11]
```diff
@@ -205,6 +215,48 @@ void InplaceMKLDNNSubgraph(std::shared_ptr<Graph> graph) {
...
+// given input output tensors
+Operation createUnaryOp(
+    std::function<void(at::Tensor output, at::Tensor input)> aten_op,
+    bool inplace = false) {
+  return [aten_op, inplace](Stack* stack) {
+    auto a = pop(stack).toTensor();
+    c10::impl::ExcludeDispatchKeyGuard edkg(c10::autograd_dispatch_keyset);
+    // we cast `a` to an `ideep::tensor`, so we can get at its descriptor
+    // which we then use to set up `out` tensor w/ the same props as a
+    auto a_it = at::native::itensor_from_mkldnn(a);
```
Comment (eellison) [inline torch/csrc/jit/passes/frozen_ops_to_mkldnn.cpp:228]: I'm going to suggest a bunch of things here for readability just because this is IMO very tricky code that is easy to get wrong (we almost did), and which requires an understanding of both mkldnn packed format and aten tensors. Feel free to not accept the comments obviously.

nit: `a_it -> a_ideep_tensor`

File: torch/csrc/jit/passes/frozen_ops_to_mkldnn.cpp [hunk 12]
```diff
@@ -205,6 +215,48 @@ void InplaceMKLDNNSubgraph(std::shared_ptr<Graph> graph) {
...
+    auto raw_data = a_it.get_data_handle();
```
Comment (eellison) [inline torch/csrc/jit/passes/frozen_ops_to_mkldnn.cpp:229]: nit: maybe `mkldnn_raw_data_handle` ?

File: torch/csrc/jit/passes/frozen_ops_to_mkldnn.cpp [hunk 13]
```diff
@@ -205,6 +215,48 @@ void InplaceMKLDNNSubgraph(std::shared_ptr<Graph> graph) {
...
+    auto topt = a.options().layout(c10::kStrided);
```
Comment (eellison) [inline torch/csrc/jit/passes/frozen_ops_to_mkldnn.cpp:230]: can you rename `topt` ? I don't know where `t` or `opt` is coming from. Maybe: `a_options_with_strided`

File: torch/csrc/jit/passes/frozen_ops_to_mkldnn.cpp [hunk 14]
```diff
@@ -205,6 +215,48 @@ void InplaceMKLDNNSubgraph(std::shared_ptr<Graph> graph) {
...
+    // we also wrap `a` storage into an aten tensor
+    auto t = at::from_blob(raw_data, {a.numel()}, topt);
```
Comment (eellison) [inline torch/csrc/jit/passes/frozen_ops_to_mkldnn.cpp:232]: can you give `t` a more verbose name ?

File: torch/csrc/jit/passes/frozen_ops_to_mkldnn.cpp [hunk 15]
```diff
@@ -205,6 +215,48 @@ void InplaceMKLDNNSubgraph(std::shared_ptr<Graph> graph) {
...
+
+    auto out_raw_data = raw_data;
+    auto out = a;
+    if (!inplace) {
+      // `a_it.get_desc()` will allocate a tensor
+      // of the right physical size.
```
Comment (eellison) [inline torch/csrc/jit/passes/frozen_ops_to_mkldnn.cpp:238]: what does right physical size here?, see comment here about more code commenting about mkldnn tensors, how they are represented, and how we are handling them here

File: torch/csrc/jit/passes/frozen_ops_to_mkldnn.cpp [hunk 16]
```diff
@@ -205,6 +215,48 @@ void InplaceMKLDNNSubgraph(std::shared_ptr<Graph> graph) {
...
 }
 
+// aten_op is a sequence of aten ops that implement a required operation
+// given input output tensors
+Operation createUnaryOp(
+    std::function<void(at::Tensor output, at::Tensor input)> aten_op,
+    bool inplace = false) {
+  return [aten_op, inplace](Stack* stack) {
+    auto a = pop(stack).toTensor();
+    c10::impl::ExcludeDispatchKeyGuard edkg(c10::autograd_dispatch_keyset);
```
Comment (eellison) [inline torch/csrc/jit/passes/frozen_ops_to_mkldnn.cpp:225]: Can you add a description at the top of this about mkldnn padded tensors, the constraint of `eltwise_ops(0) = 0`, and what we are about to do ?

File: torch/csrc/jit/passes/frozen_ops_to_mkldnn.cpp [hunk 17]
```diff
@@ -238,6 +291,57 @@ Operation BroadOp(const Node* node) {
   };
 }
 
+const RegisterOperators MKLDNNHardSwishOpReg({
```
Comment (eellison) [inline torch/csrc/jit/passes/frozen_ops_to_mkldnn.cpp:294]: Can you add a comment about the constraint that `op(0) == 0`

File: torch/csrc/jit/passes/frozen_ops_to_mkldnn.cpp [hunk 18]
```diff
@@ -205,6 +215,48 @@ void InplaceMKLDNNSubgraph(std::shared_ptr<Graph> graph) {
...
+    auto topt = a.options().layout(c10::kStrided);
+    // we also wrap `a` storage into an aten tensor
+    auto t = at::from_blob(raw_data, {a.numel()}, topt);
+
+    auto out_raw_data = raw_data;
+    auto out = a;
+    if (!inplace) {
+      // `a_it.get_desc()` will allocate a tensor
+      // of the right physical size.
+      auto it_empty = ideep::tensor(a_it.get_desc());
```
Comment (eellison) [inline torch/csrc/jit/passes/frozen_ops_to_mkldnn.cpp:239]: maybe add an assertion that it_empty physical size equals input physical size?

File: torch/csrc/jit/passes/frozen_ops_to_mkldnn.cpp [hunk 19]
```diff
@@ -456,11 +560,23 @@ void ComputeSubgraphInMKLDNN(Node* subgraph_node) {
       body_node->replaceInput(1, node->outputs().at(1));
     }
 
-    if (body_node->kind() == aten::conv2d ||
-        body_node->kind() == aten::conv3d) {
-      // this node doesnt handle string padding yet...
-      if (!body_node->namedInput("padding")->type()->cast<StringType>()) {
-        body_node->replaceWithNewSymbol(Symbol::prim("mkldnn_convolution"));
+    auto true_pred = [](Node*) { return true; };
```
Comment (eellison) [inline torch/csrc/jit/passes/frozen_ops_to_mkldnn.cpp:563]: Nit: can you keep the conv handling how it was originally? I think it was more readable originally and we've lost the `this node doesnt handle string padding yet` comment. If we start getting a lot more predicates than maybe we should refactor.

File: torch/csrc/jit/passes/frozen_ops_to_mkldnn.cpp [hunk 20]
```diff
@@ -238,6 +291,57 @@ Operation BroadOp(const Node* node) {
...
+            [](at::Tensor output, at::Tensor input) {
+              at::hardsigmoid_out(output, input);
+            },
+            true),
+        AliasAnalysisKind::FROM_SCHEMA),
+    torch::jit::Operator(
+        "prim::MKLDNNRelu6_(Tensor(a!) self) -> Tensor(a!)",
+        createUnaryOp(
+            [](at::Tensor output, at::Tensor input) {
+              at::hardtanh_out(output, input, 0.f, 6.f);
```
Comment (eellison) [inline torch/csrc/jit/passes/frozen_ops_to_mkldnn.cpp:315]: Same comment here, just use relu6 since it exists

File: torch/csrc/jit/passes/frozen_ops_to_mkldnn.cpp [hunk 21]
```diff
@@ -238,6 +291,57 @@ Operation BroadOp(const Node* node) {
...
+            [](at::Tensor output, at::Tensor input) {
+              at::hardsigmoid_out(output, input);
+            },
+            false),
+        AliasAnalysisKind::FROM_SCHEMA),
+    torch::jit::Operator(
+        "prim::MKLDNNRelu6(Tensor(a!) self) -> Tensor(a!)",
+        createUnaryOp(
+            [](at::Tensor output, at::Tensor input) {
+              at::hardtanh_out(output, input, 0.f, 6.f);
```
Comment (eellison) [inline torch/csrc/jit/passes/frozen_ops_to_mkldnn.cpp:339]: There is an `at::relu6`, can you use it instead of `hardtanh`?

File: torch/csrc/jit/passes/frozen_ops_to_mkldnn.cpp [hunk 22]
```diff
@@ -238,6 +291,57 @@ Operation BroadOp(const Node* node) {
   };
 }
 
+const RegisterOperators MKLDNNHardSwishOpReg({
+    torch::jit::Operator(
+        "prim::MKLDNNHardSwish_(Tensor(a!) self) -> Tensor(a!)",
+        createUnaryOp(
+            [](at::Tensor output, at::Tensor input) {
+              at::hardswish_out(output, input);
```
Comment (eellison) [inline torch/csrc/jit/passes/frozen_ops_to_mkldnn.cpp:299]: For these operators, we could just call `at::cpu::hardswish_out`. Since we are already creating a custom operator we might as well avoid the overhead

**Review by eellison — COMMENTED** [review]
This looks good and I think is right (great job, this was tricky) but can you add a little more testing and more of a comment ? Reading through the code right now I wouldn't know what invariants we have to maintain and why the code is written how it is

File: test/jit/test_freezing.py [hunk 23]
```diff
@@ -1825,6 +1827,34 @@ def forward(self, x):
...
+                        graph_str = f"""graph(%input.1 : Tensor):
+                            %33 : None = prim::Constant()
+                            %34 : Tensor = aten::to_mkldnn(%input.1, %33)
+                            %35 : Tensor = {mkldnn_opname}{inplace_str}(%34)
+                            return ({inplace_tgt})
+                        """
+                        g = parse_ir(graph_str)
+                        m = self.createFunctionFromGraph(g)
+                        x = torch.rand(size)
+                        # `inplace=False` is intentional, otherwise we modify the input
```
[the remark this answers is not in the document]
Comment (Krovatkin) [reply test/jit/test_freezing.py:1874]: added a conv test and we already have inplace tests.

[the remark this answers is not in the document]
Comment (eellison) [reply test/jit/test_freezing.py:1874]: we have inplace tests that arent testing these new ops which have a different way of running them
